orangecontrib/prototypes/widgets/contingency_table.py

```diff
@@ -0,0 +1,312 @@
+from math import isnan, isinf
+
+import unicodedata
+from AnyQt.QtCore import Qt, QItemSelection, QItemSelectionModel
+from AnyQt.QtGui import QStandardItem, QColor, QFont, QBrush
+from AnyQt.QtWidgets import QTableView, QSizePolicy, QHeaderView, QStyledItemDelegate
+from Orange.widgets import gui
+
+
+BorderRole = next(gui.OrangeUserRole)
+BorderColorRole = next(gui.OrangeUserRole)
+
+
+class BorderedItemDelegate(QStyledItemDelegate):
+    """Item delegate that paints border at the specified sides
+
+    Data for `BorderRole` is a string containing letters t, r, b and/or l,
+    which defines the sides at which the border is drawn.
+
+    Role `BorderColorRole` sets the color for the cell. If not color is given,
+    `self.color` is used as default.
+
+    Args:
+        color (QColor): default color (default default is black)
+    """
+    def __init__(self, color=Qt.black):
+        super().__init__()
+        self.color = color
+
+    def paint(self, painter, option, index):
+        """Overloads `paint` to draw borders"""
+        QStyledItemDelegate.paint(self, painter, option, index)
+        borders = index.data(BorderRole)
+        if borders:
+            color = index.data(BorderColorRole) or self.color
+            painter.save()
+            painter.setPen(color)
+            rect = option.rect
+            for side, p1, p2 in (("t", rect.topLeft(), rect.topRight()),
+                                 ("r", rect.topRight(), rect.bottomRight()),
+                                 ("b", rect.bottomLeft(), rect.bottomRight()),
+                                 ("l", rect.topLeft(), rect.bottomLeft())):
+                if side in borders:
+                    painter.drawLine(p1, p2)
+            painter.restore()
+
+
+class ContingencyTable(QTableView):
+    """
+    A contingency table widget which can be used wherever ``QTableView`` could be used.
+
+    Parameters
+    ----------
+    parent : Orange.widgets.widget.OWWidget as a
+        The containing widget to which the table is connected.
+    tablemodel : AnyQt.QtGui.QtStandardItemModel
+        The model in which table data will be stored.
+
+    Attributes
+    ----------
+    classesv : :obj:`list` of :obj:`str`
+        Vertical class headers.
+    classesh : :obj:`list` of :obj:`str`
+        Horizontal class headers.
+    headerv : :obj:`str`, optional
+        Vertical top header.
+    headerh : :obj:`str`, optional
+        Horizontal top header.
+    corner_string : str
+        String that is top right and bottom left corner of the table.
+        Default is ``unicodedata.lookup("N-ARY SUMMATION")``.
+    """
+
+    def __init__(self, parent, tablemodel):
+        super().__init__(editTriggers=QTableView.NoEditTriggers)
+
+        self.classesv = None
+        self.classesh = None
+        self.headerv = None
+        self.headerh = None
+        self.parent = parent
+        self.tablemodel = tablemodel
```

Comment:
As discused in comments for OWContingencyTable.

``` 
self.tablemodel = QStandardItemModel() 
```

Reply:
And for clearing data
```
def clear(self):
    self.tablemodel.clear()
```

Reply:
I agree with @JakaKokosar on this. ContingencyTable should be in charge of creating the model as it already fills it with custom items.